[Console] Support suggesting index templates v2

mibragimov opened the pull request:
Closes https://github.com/elastic/kibana/issues/75967

### Release note
Console in Dev Tools now supports autocompletion for [index templates](https://www.elastic.co/guide/en/elasticsearch/reference/master/index-templates.html) and [component templates](https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-component-template.html) introduced in Elasticsearch 7.8.

<img width="544" alt="Screen Shot 2022-02-09 at 9 31 29 AM" src="https://user-images.githubusercontent.com/5226211/153222113-c0bdf3b0-15b4-401b-b005-93d3de193de1.png">

![Capture](https://user-images.githubusercontent.com/53621505/153364505-65795879-fb73-4644-a039-5812b0e025ed.PNG)

Kunzetsov commented on src/plugins/console/public/lib/kb/kb.js:
```diff
@@ -86,6 +87,9 @@ const parametrizedComponentFactories = {
   node: function (name, parent) {
     return new ListComponent(name, [], parent, false);
   },
+  name: function (name, parent) {
```
Definitely, should be changed to the `index_template`.

Kunzetsov commented on src/plugins/console/public/lib/mappings/mappings.js:
```diff
@@ -186,6 +191,10 @@ function loadTemplates(templatesObject = {}) {
 }
 
+function loadIndexTemplates(data) {
+  indexTemplates = _.map(data.index_templates, 'name');
```
Should be changed to:
```js
indexTemplates = (data.index_templates ?? []).map(({ name }) => name);
```

It will be painful to track the changes of all the string fields, used at `get`/`set`/`map`/etc, even after moving to TS, and it is better to avoid such a practice where it is possible. This code can be easily converted to vanilla js.

Kunzetsov commented on src/plugins/console/public/lib/mappings/mappings.js:
```diff
@@ -292,42 +303,52 @@ export function retrieveAutoCompleteInfo(settings, settingsToRetrieve) {
+    (mappings, aliases, templates, indexTemplates) => {
+      let mappingsResponse;
+      if (mappings) {
+        const maxMappingSize = mappings[0].length > 10 * 1024 * 1024;
```
I know, this code is not yours, but, please, think about this code. It can fail, if `mappings` is an empty array.

And even more, this code is failing silently. We should wrap it up with `try/catch` and provide some notifying about the error. Could you, please, add some correct checks for this code to be safe during the execution? Thanks.

Kunzetsov reviewed (commented):
Solution LGTM. Thanks, @mibragimov. 
Left some comments to be addressed.

Kunzetsov commented:
@elasticmachine merge upstream

mibragimov commented on src/plugins/console/public/lib/autocomplete/components/index_template_autocomplete_component.js:
```diff
@@ -0,0 +1,24 @@
+  }
+
+  getDefaultTermMeta() {
+    return 'name';
```
There is bit of a problem with long index template names. See screenshots below:

![Capture1](https://user-images.githubusercontent.com/53621505/152925532-82357d88-a2e3-43df-928b-863801042147.PNG)
![Capture](https://user-images.githubusercontent.com/53621505/152925519-1b99144c-afc4-4a08-950f-5a907c48cf31.PNG)

Perhaps we should leave `name` as default here. WDYT?

Kunzetsov replied:
The main idea of the context key is to identify the element, which has been shown at the prompt. Let us imagine, one day we'll receive the task to add templates to _index_template's autocompletion results.

And you'll have: 
```
.ml-state-temp                       template
.ml-state                            name
```

But should have 
```
.ml-state-temp                       template
.ml-state                            index_template
```

What about the long name of the index, possibly, we should pick some shorter meaningful name for the index_template context key or, as such a problem has been before, leave it as it is.

mibragimov replied:
@cjcenizal could you please review the changes and advise regarding the comments above?

cjcenizal replied:
Thanks for the ping! Someone from @elastic/platform-deployment-management will provide input.

alisonelizabeth replied:
I agree with @mibragimov and would prefer to use `index_template`.

I see the issue with the spacing, though. Perhaps we should address this as a separate issue, as the same could happen in other scenarios, e.g., if you have a really long index name.

<img width="326" alt="Screen Shot 2022-02-09 at 9 26 33 AM" src="https://user-images.githubusercontent.com/5226211/153221535-3a9bfd40-fbb6-4b4b-8b13-6ba7b338f9ea.png">

mibragimov replied:
Created a separate issue for this https://github.com/elastic/kibana/issues/125186. I'll work on that

Kunzetsov reviewed (approved):
Code LGTM 👍 

What about the width of the autocomplete prompt window, let us listen to the opinion of the @elastic/platform-deployment-management.

elasticmachine commented:
Pinging @elastic/platform-deployment-management (Team:Deployment Management)

alisonelizabeth commented:
@elasticmachine merge upstream

alisonelizabeth reviewed (commented):
Great work @mibragimov!

Tested locally and works as expected. I went ahead and updated the PR description to include a release note. I also updated the labels to include `8.2`. I believe `main` is currently representative of `8.2`, and then this will need to be backported to the `8.1` branch.

Also, a couple questions - 

1 - Is it possible to add any tests around this new functionality?
2 - What are your thoughts about renaming areas of the code for the legacy template autocomplete to be prefixed with `legacy`? For example, `TemplateAutocompleteComponent` would become `LegacyTemplateAutocompleteComponent`. Otherwise, I could see it potentially confusing if someone else comes into the code and needs to differentiate between the template autocomplete logic vs. the index_template autocomplete logic. WDYT?

mibragimov commented:
> Great work @mibragimov!
> 
> Tested locally and works as expected. I went ahead and updated the PR description to include a release note. I also updated the labels to include `8.2`. I believe `main` is currently representative of `8.2`, and then this will need to be backported to the `8.1` branch.
> 
> Also, a couple questions -
> 
> 1 - Is it possible to add any tests around this new functionality? 2 - What are your thoughts about renaming areas of the code for the legacy template autocomplete to be prefixed with `legacy`? For example, `TemplateAutocompleteComponent` would become `LegacyTemplateAutocompleteComponent`. Otherwise, I could see it potentially confusing if someone else comes into the code and needs to differentiate between the template autocomplete logic vs. the index_template autocomplete logic. WDYT?

Thank you for the review @alisonelizabeth! I completely agree with the suggested changes. Working on these

mibragimov commented:
@alisonelizabeth @Kunzetsov  I haven't noticed this point from https://github.com/elastic/kibana/issues/75967

> Note that we also want to suggest the names of existing component templates for the `_component_template` API.

So I added support for `_component_template` as well. Working on the tests.

![Capture](https://user-images.githubusercontent.com/53621505/153363664-d10c02f0-4b1f-4eb1-b38b-46d32ca54bed.PNG)

alisonelizabeth commented:
Thanks for making those changes @mibragimov! Let me know when you add tests and I will review again.

mibragimov commented:
@elasticmachine merge upstream

alisonelizabeth reviewed (approved):
Latest changes LGTM. Thanks @mibragimov!

kibanamachine commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.1|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/125612">](https://github.com/elastic/kibana/pull/125612)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sqren/backport)
